From 3ef66b05ed0ec19288ebaf4172e73c0b52e20d5f Mon Sep 17 00:00:00 2001
From: Tianfei zhang <Tianfei.zhang@intel.com>
Date: Thu, 13 Jan 2022 16:07:34 +0800
Subject: [PATCH 70/93] fpga: m10bmc-sec: PR/SR root key hash support for VAB

Add support for reading the Secure Device Manager (SDM) PR/SR
Root Entry Hash and Cancellation Key. This support is required
for Vender Authenticated Boot (VAB).

Signed-off-by: Tianfei zhang <Tianfei.zhang@intel.com>
---
 .../sysfs-driver-intel-m10-bmc-sec-update     | 37 +++++++++
 drivers/fpga/intel-m10-bmc-sec-update.c       | 76 ++++++++++++++++++-
 drivers/mfd/intel-m10-bmc-core.c              |  4 +
 include/linux/mfd/intel-m10-bmc.h             | 13 ++++
 4 files changed, 129 insertions(+), 1 deletion(-)

diff --git a/Documentation/ABI/testing/sysfs-driver-intel-m10-bmc-sec-update b/Documentation/ABI/testing/sysfs-driver-intel-m10-bmc-sec-update
index 0872d80bb1ee..b516436b2889 100644
--- a/Documentation/ABI/testing/sysfs-driver-intel-m10-bmc-sec-update
+++ b/Documentation/ABI/testing/sysfs-driver-intel-m10-bmc-sec-update
@@ -148,3 +148,40 @@ Contact:	Tianfei zhang <tianfei.zhang@intel.com>
 Description:	Read-only. Reading this file will return the name of image booted
 		from FPGA. The EINVAL error code will be returned if no image booted
 		from FPGA.
+
+What:		/sys/bus/platform/drivers/intel-m10bmc-sec-update/.../security/sr_sdm_root_entry_hash
+Date:		Jan 2022
+KernelVersion:	5.16
+Contact:	Tianfei Zhang <tianfei.zhang@intel.com>
+Description:	Read only. Returns the root entry hash of SDM (Secure Device
+		Manager) for the static region if one is programmed, else it
+		returns the string: "hash not programmed".  This file is only
+		visible if the underlying device supports it.
+		Format: "0x%x".
+
+What:		/sys/bus/platform/drivers/intel-m10bmc-sec-update/.../security/pr_sdm_root_entry_hash
+Date:		Jan 2022
+KernelVersion:	5.16
+Contact:	Tianfei Zhang <tianfei.zhang@intel.com>
+Description:	Read only. Returns the root entry hash of SDM (Secure Device
+		Manager) for the partial reconfiguration region if one is programmed,
+		else it returns the string: "hash not programmed".  This file
+		is only visible if the underlying device supports it.
+		Format: "0x%x".
+
+What:		/sys/bus/platform/drivers/intel-m10bmc-sec-update/.../security/pr_sdm_canceled_csks
+Date:		Jan 2022
+KernelVersion:	5.16
+Contact:	Tianfei Zhang <tianfei.zhang@intel.com>
+Description:	Read only. Returns a list of indices for canceled code
+		signing keys of SDM (Secure Device Manager) for the partial
+		reconfiguration region. The standard bitmap list format is
+		used (e.g. "1,2-6,9").
+
+What:		/sys/bus/platform/drivers/intel-m10bmc-sec-update/.../security/sr_sdm_canceled_csks
+Date:		Jan 2022
+KernelVersion:	5.16
+Contact:	Tianfei Zhang <tianfei.zhang@intel.com>
+Description:	Read only. Returns a list of indices for canceled code
+		signing keys of SDM (Secure Device Manager) for the static
+		region. The standard bitmap list format is used (e.g. "1,2-6,9").
diff --git a/drivers/fpga/intel-m10-bmc-sec-update.c b/drivers/fpga/intel-m10-bmc-sec-update.c
index 0e1b9bff8da2..5492c1e68cab 100644
--- a/drivers/fpga/intel-m10-bmc-sec-update.c
+++ b/drivers/fpga/intel-m10-bmc-sec-update.c
@@ -533,6 +533,44 @@ DEVICE_ATTR_SEC_REH_RO(bmc);
 DEVICE_ATTR_SEC_REH_RO(sr);
 DEVICE_ATTR_SEC_REH_RO(pr);
 
+#define SDM_ROOT_HASH_REG_NUM 12
+
+static ssize_t
+show_sdm_root_entry_hash(struct device *dev, u32 start, char *buf)
+{
+	struct m10bmc_sec *sec = dev_get_drvdata(dev);
+	int i, cnt, ret;
+	u32 key;
+
+	cnt = sprintf(buf, "0x");
+	for (i = 0; i < SDM_ROOT_HASH_REG_NUM; i++) {
+		ret = m10bmc_sys_read(sec->m10bmc,
+				      m10bmc_base(sec->m10bmc) +
+				      start + i * 4, &key);
+		if (ret)
+			return ret;
+
+		cnt += sprintf(buf + cnt, "%08x", key);
+	}
+	cnt += sprintf(buf + cnt, "\n");
+
+	return cnt;
+}
+
+#define DEVICE_ATTR_SDM_SEC_REH_RO(_name) \
+static ssize_t _name##_sdm_root_entry_hash_show(struct device *dev, \
+					    struct device_attribute *attr, \
+					    char *buf) \
+{							\
+	struct m10bmc_sec *sec = dev_get_drvdata(dev);   \
+	struct intel_m10bmc *m10bmc = sec->m10bmc;  \
+	return show_sdm_root_entry_hash(dev, _name##_sdm_reh_reg(m10bmc),\
+			buf); } \
+static DEVICE_ATTR_RO(_name##_sdm_root_entry_hash)
+
+DEVICE_ATTR_SDM_SEC_REH_RO(pr);
+DEVICE_ATTR_SDM_SEC_REH_RO(sr);
+
 #define CSK_BIT_LEN		128U
 #define CSK_32ARRAY_SIZE	DIV_ROUND_UP(CSK_BIT_LEN, 32)
 
@@ -577,6 +615,34 @@ DEVICE_ATTR_SEC_CSK_RO(bmc);
 DEVICE_ATTR_SEC_CSK_RO(sr);
 DEVICE_ATTR_SEC_CSK_RO(pr);
 
+static ssize_t
+show_sdm_canceled_csk(struct device *dev, u32 addr, char *buf)
+{
+	struct m10bmc_sec *sec = dev_get_drvdata(dev);
+	int ret;
+	u32 val;
+
+	ret = m10bmc_sys_read(sec->m10bmc,
+			      m10bmc_base(sec->m10bmc) + addr, &val);
+	if (ret)
+		return ret;
+
+	return sysfs_emit(buf, "%08x\n", val);
+}
+
+#define DEVICE_ATTR_SDM_SEC_CSK_RO(_name) \
+static ssize_t _name##_sdm_canceled_csks_show(struct device *dev, \
+					  struct device_attribute *attr, \
+					  char *buf) \
+{                                                    \
+	struct m10bmc_sec *sec = dev_get_drvdata(dev);   \
+	struct intel_m10bmc *m10bmc = sec->m10bmc;  \
+	return show_sdm_canceled_csk(dev, _name##_sdm_csk_reg(m10bmc),\
+			buf); } \
+static DEVICE_ATTR_RO(_name##_sdm_canceled_csks)
+DEVICE_ATTR_SDM_SEC_CSK_RO(pr);
+DEVICE_ATTR_SDM_SEC_CSK_RO(sr);
+
 #define FLASH_COUNT_SIZE 4096	/* count stored as inverted bit vector */
 
 static ssize_t flash_count_show(struct device *dev,
@@ -690,7 +756,11 @@ m10bmc_security_is_visible(struct kobject *kobj, struct attribute *attr, int n)
 	    (attr == &dev_attr_sdm_sr_provision_status.attr ||
 	     attr == &dev_attr_sdm_sr_cancel_status.attr ||
 	     attr == &dev_attr_sdm_pr_provision_status.attr ||
-	     attr == &dev_attr_sdm_pr_cancel_status.attr))
+	     attr == &dev_attr_sdm_pr_cancel_status.attr ||
+	     attr == &dev_attr_pr_sdm_root_entry_hash.attr ||
+	     attr == &dev_attr_pr_sdm_canceled_csks.attr ||
+	     attr == &dev_attr_sr_sdm_root_entry_hash.attr ||
+	     attr == &dev_attr_sr_sdm_canceled_csks.attr))
 		return 0;
 
 	return attr->mode;
@@ -708,6 +778,10 @@ static struct attribute *m10bmc_security_attrs[] = {
 	&dev_attr_sdm_sr_cancel_status.attr,
 	&dev_attr_sdm_pr_provision_status.attr,
 	&dev_attr_sdm_pr_cancel_status.attr,
+	&dev_attr_pr_sdm_root_entry_hash.attr,
+	&dev_attr_pr_sdm_canceled_csks.attr,
+	&dev_attr_sr_sdm_root_entry_hash.attr,
+	&dev_attr_sr_sdm_canceled_csks.attr,
 	NULL,
 };
 
diff --git a/drivers/mfd/intel-m10-bmc-core.c b/drivers/mfd/intel-m10-bmc-core.c
index a84cc0ca1908..7efea9965a25 100644
--- a/drivers/mfd/intel-m10-bmc-core.c
+++ b/drivers/mfd/intel-m10-bmc-core.c
@@ -29,6 +29,10 @@ static const struct m10bmc_csr m10bmc_pmci_csr = {
 	.pr_reh_addr = PMCI_PR_REH_ADDR,
 	.pr_magic = PMCI_PR_PROG_MAGIC,
 	.rsu_update_counter = PMCI_STAGING_FLASH_COUNT,
+	.pr_sdm_reh_reg = M10BMC_PMCI_PR_RH0,
+	.pr_sdm_csk_reg = M10BMC_PMCI_PR_CSK,
+	.sr_sdm_reh_reg = M10BMC_PMCI_SR_RH0,
+	.sr_sdm_csk_reg = M10BMC_PMCI_SR_CSK,
 };
 
 static const struct m10bmc_csr m10bmc_spi_csr = {
diff --git a/include/linux/mfd/intel-m10-bmc.h b/include/linux/mfd/intel-m10-bmc.h
index 9a1a607c7243..7f29bbb41f21 100644
--- a/include/linux/mfd/intel-m10-bmc.h
+++ b/include/linux/mfd/intel-m10-bmc.h
@@ -276,6 +276,11 @@ enum m10bmc_type {
 #define M10BMC_PMCI_CERT_PROG_STS	0x824
 #define M10BMC_PMCI_CERT_SPEC_STS	0x828
 
+#define M10BMC_PMCI_SR_RH0 0x848
+#define M10BMC_PMCI_SR_CSK 0x878
+#define M10BMC_PMCI_PR_RH0 0x87c
+#define M10BMC_PMCI_PR_CSK 0x8ac
+
 #define PMCI_ERROR_LOG_ADDR  0x7fb0000
 #define PMCI_ERROR_LOG_SIZE  0x40000
 
@@ -314,6 +319,10 @@ enum m10bmc_type {
 #define pr_reh_addr(m10bmc) ((m10bmc)->csr->pr_reh_addr)
 #define pr_magic(m10bmc) ((m10bmc)->csr->pr_magic)
 #define rsu_update_counter(m10bmc) ((m10bmc)->csr->rsu_update_counter)
+#define pr_sdm_reh_reg(m10bmc) ((m10bmc)->csr->pr_sdm_reh_reg)
+#define pr_sdm_csk_reg(m9bmc) ((m10bmc)->csr->pr_sdm_csk_reg)
+#define sr_sdm_reh_reg(m10bmc) ((m10bmc)->csr->sr_sdm_reh_reg)
+#define sr_sdm_csk_reg(m10bmc) ((m10bmc)->csr->sr_sdm_csk_reg)
 
 enum m10bmc_fw_state {
 	M10BMC_FW_STATE_NORMAL,
@@ -341,6 +350,10 @@ struct m10bmc_csr {
 	unsigned int pr_reh_addr;
 	unsigned int pr_magic;
 	unsigned int rsu_update_counter;
+	unsigned int pr_sdm_reh_reg;
+	unsigned int pr_sdm_csk_reg;
+	unsigned int sr_sdm_reh_reg;
+	unsigned int sr_sdm_csk_reg;
 };
 
 /**
-- 
2.39.0

